tests/common/backend.py

```diff
@@ -555,6 +565,109 @@ def correct_addr(self, addr: BackendAddr) -> BackendAddr:
     async def connection_factory(self) -> trio.abc.Stream:
         return await trio.open_tcp_stream(self.addr.hostname, self.addr.port)
 
+    def test_client(self, use_cookies: bool = True) -> TestClientProtocol:
+        """Creates and returns a test client"""
+        return self.asgi_app.test_client(use_cookies)
+
+
+class AuthenticatedHttpApiClient:
```

Comment:
Replace this by the class I suggested in https://github.com/Scille/parsec-cloud/pull/2544#discussion_r955070416

this has the added value of turning this:
```python
@pytest.mark.trio
async def test_signature_missing_header(alice_http_client: AuthenticatedHttpApiClient):
   timestamp = DateTime.now()

    request = alice_http_client.get_request()

    headers = alice_http_client.build_authenticated_headers(
        api_version=API_VERSION,
        author=alice_http_client.device_id,
        timestamp=timestamp,
    )

    rep = await alice_http_client.send_request(headers, request)
    assert rep.status_code == 400  # bad request
```

into this:

```python
@pytest.mark.trio
async def test_signature_missing_header(alice_http_client: AuthenticatedHttpApiClient):
    rep = await alice_http_client.send_ping(extra_headers={"Signature": None})
    assert rep.status_code == 400  # bad request
```

Reply:
For more advanced customization of the headers, we should add special params to the send:

```python
class AuthenticatedHttpApiClient:
    [...]
    # now and before_send_hook for easy customization
    async def send(self, req, serializer, extra_headers: Dict[str, str]= {}, before_send_hook=None, now=None, check_rep: bool = True):
        now = now or DateTime.now()
        body = serializer.req_dumps(req)
        headers = {
            **self.base_headers,
            "Timestamp": now.to_rfc(),
            "Signature": signing_request(self.device.signing_key, self.device.device_id, now, body)
        }

        # Customize headers
        for k, v in extra_headers.items():
            if v is None:
                headers.pop(k, None)
            else:
                headers[k] = v

        args = {"url": f"/authenticated/{self.device.organization_id}", "headers": headers, "data": body}
        # Last chance to customize the request !
        if before_send_hook:
            # Passing as dict allow the hook to event modify the url param
            before_send_hook(args)
        rep = await self.client.post(**args)

        if check_rep:
            assert rep.status_code == 200
            rep_body = await rep.body()
            return serializer.rep_loads(rep_body)

        else:
            return rep
```
then

```python
@pytest.mark.trio
async def test_signature_invalid(alice_http_client: AuthenticatedHttpApiClient):
    timestamp = DateTime.now()

    request = alice_http_client.get_request()

    signature = alice_http_client.sign_request(timestamp.add(minutes=1), request)

    headers = alice_http_client.build_authenticated_headers(
        api_version=API_VERSION,
        author=alice_http_client.device_id,
        timestamp=timestamp,
        signature=signature,
    )

    rep = await alice_http_client.send_request(headers, request)
    assert rep.status_code == 401  # Unauthorized
```

becomes

```python
@pytest.mark.trio
async def test_signature_invalid(alice_http_client: AuthenticatedHttpApiClient):
    def _before_send_hook(args):
        args["headers"]["signature"] = "dummy"
    alice_http_client.send_ping(before_send_hook=_before_send_hook)
    assert rep.status_code == 401  # Unauthorized

# Or with a valid but in the future signature:
@pytest.mark.trio
async def test_signature_invalid(alice_http_client: AuthenticatedHttpApiClient):
    alice_http_client.send_ping(now=DateTime.now().add(minutes=1))
    assert rep.status_code == 401  # Unauthorized
```

Reply:
done